### PR TITLE
Refactor the subsampling module to unify the behaviour of equilibrium_detection and statistical_inefficiency

### PR DESCRIPTION
**Pull request by xiki-tempula**
- Add the keyword `remove_burnin` to the high-level function `decorrelate_u_nk` and `decorrelate_dhdl` to use `equilibrium_detection` instead of `statistical_inefficiency`.
- Extract the function `u_nk2series` and `dhdl2series` to make series from dataframe
- Extract the function `_drop_duplicates`, `_sort_by_time` and `_prepare_input` to clean the data before passing to `equilibrium_detection` and `statistical_inefficiency`
- Unify the behaviour of `equilibrium_detection` and `statistical_inefficiency` in terms of preprocessing and checking the data.

### CODE REVIEW

**Comment by orbeckst** [comment]
Looks neat — should be good for 1.0.0. Will definitely need CHANGELOG entries and `versionchanged/added 1.0.0` markers.

I am not sure if I get to review this over the next 10 days — busy + ACS conference. If anyone else can review then that would be fabulous.

File: CHANGES [hunk 1]
```diff
@@ -21,6 +21,8 @@ Changes
   - Drop support for py3.7 (Issue #179, PR #214)
 
 Enhancements
+  - Refactor the subsampling module to add u_nk2series and dhdl2series to
+  convert u_nk and dHdl to series. (PR #218)
```
Comment (orbeckst) [inline CHANGES:25]: indentation
```suggestion
    convert u_nk and dHdl to series. (PR #218)
```

File: docs/preprocessing/alchemlyb.preprocessing.subsampling.rst [hunk 2]
```diff
@@ -5,10 +5,58 @@ Subsampling
...
-The functions featured in this module can be used to easily subsample either :ref:`dHdl <dHdl>` or :ref:`u_nk <u_nk>` datasets to give less correlated timeseries.
+The functions featured in this module can be used to easily subsample either
+:ref:`dHdl <dHdl>` or :ref:`u_nk <u_nk>` datasets to give less correlated
+timeseries.
+
+High-level functions
+--------------------
+Two high-level functions
+:func:`~alchemlyb.preprocessing.subsampling.decorrelate_u_nk` and
+:func:`~alchemlyb.preprocessing.subsampling.decorrelate_dhdl` could be used to
```
Comment (orbeckst) [inline docs/preprocessing/alchemlyb.preprocessing.subsampling.rst:16]:
```suggestion
:func:`~alchemlyb.preprocessing.subsampling.decorrelate_dhdl` can be used to
```

File: docs/preprocessing/alchemlyb.preprocessing.subsampling.rst [hunk 3]
```diff
@@ -5,10 +5,58 @@ Subsampling
...
+preprocess the :ref:`dHdl <dHdl>` or :ref:`u_nk <u_nk>` in a automatic
```
Comment (orbeckst) [inline docs/preprocessing/alchemlyb.preprocessing.subsampling.rst:17]:
```suggestion
preprocess the :ref:`dHdl <dHdl>` or :ref:`u_nk <u_nk>` in an automatic
```

File: docs/preprocessing/alchemlyb.preprocessing.subsampling.rst [hunk 4]
```diff
@@ -5,10 +5,58 @@ Subsampling
...
+fashion. The following code could remove the initial burnin period and
+decorrelate the data. ::
```
Comment (orbeckst) [inline docs/preprocessing/alchemlyb.preprocessing.subsampling.rst:19]:
```suggestion
fashion. The following code removes an initial "burnin" period and
decorrelates the data. ::
```

File: src/alchemlyb/preprocessing/subsampling.py [hunk 5]
```diff
@@ -29,25 +27,111 @@ def decorrelate_u_nk(df, method='dhdl', drop_duplicates=True,
...
+    Returns
+    -------
+    Series
+        `series` to be used as input for
+        :func:`~alchemlyb.preprocessing.subsampling.statistical_inefficiency`
+        or
+        :func:`~alchemlyb.preprocessing.subsampling.equilibrium_detection`.
+
+
+    .. versionadded:: 0.7.0
```
Comment (orbeckst) [inline src/alchemlyb/preprocessing/subsampling.py:133]: 1.0.0

File: src/alchemlyb/preprocessing/subsampling.py [hunk 6]
```diff
@@ -29,25 +27,111 @@ def decorrelate_u_nk(df, method='dhdl', drop_duplicates=True,
...
+    DataFrame
+        `df` subsampled.
     Note
     ----
     The default of ``True`` for  `drop_duplicates` and `sort` should result in robust decorrelation
     but can loose data.
 
+
     .. versionadded:: 0.6.0
+    .. versionchanged:: 0.7.0
```
Comment (orbeckst) [inline src/alchemlyb/preprocessing/subsampling.py:94]: 1.0.0

File: src/alchemlyb/preprocessing/subsampling.py [hunk 7]
```diff
@@ -84,49 +168,36 @@ def decorrelate_u_nk(df, method='dhdl', drop_duplicates=True,
...
-    but can loose data.
+    Series
+        `series` to be used as input for
+        :func:`~alchemlyb.preprocessing.subsampling.statistical_inefficiency`
+        or
+        :func:`~alchemlyb.preprocessing.subsampling.equilibrium_detection`.
 
 
-    .. versionadded:: 0.6.0
+    .. versionadded:: 0.7.0
```
Comment (orbeckst) [inline src/alchemlyb/preprocessing/subsampling.py:196]: 1.0.0

File: src/alchemlyb/preprocessing/subsampling.py [hunk 8]
```diff
@@ -344,30 +488,28 @@ def equilibrium_detection(df, series=None, lower=None, upper=None, step=None):
...
-    """
-    if _check_multiple_times(df):
-        raise KeyError("Duplicate time values found; equilibrium detection "
-                       "is only meaningful for a single, contiguous, "
-                       "and sorted timeseries.")
 
-    if not _check_sorted(df):
-        raise KeyError("Equilibrium detection only works as expected if "
-                       "values are sorted by time, increasing.")
+    .. versionchanged:: 0.7.0
```
Comment (orbeckst) [inline src/alchemlyb/preprocessing/subsampling.py:494]: 1.0.0

Comment (orbeckst) [reply src/alchemlyb/preprocessing/subsampling.py:494]: update CHANGES

File: src/alchemlyb/preprocessing/subsampling.py [hunk 9]
```diff
@@ -344,30 +488,28 @@ def equilibrium_detection(df, series=None, lower=None, upper=None, step=None):
...
+       Add the drop_duplicates and sort keyword to unify the behaviour between
+       equilibrium_detection and statistical_inefficiency.
```
Comment (orbeckst) [inline src/alchemlyb/preprocessing/subsampling.py:496]: Use markup to make it look nicer in the online docs.

File: src/alchemlyb/preprocessing/subsampling.py [hunk 10]
```diff
@@ -29,25 +27,111 @@ def decorrelate_u_nk(df, method='dhdl', drop_duplicates=True,
...
     DataFrame
         `df` subsampled according to selected `method`.
+    Note
+    ----
+    The default of ``True`` for  `drop_duplicates` and `sort` should result in robust decorrelation
+    but can loose data.
+
 
+    .. versionadded:: 0.6.0
+    .. versionchanged:: 0.7.0
```
Comment (orbeckst) [inline src/alchemlyb/preprocessing/subsampling.py:47]: 1.0.0

**Review by orbeckst — CHANGES_REQUESTED** [review]
Looks very good, just some small typos, update the release number in the docs, and document everything in CHANGES.

**Comment by orbeckst** [comment]
Can you check why RTD fails?

**Comment by xiki-tempula** [comment]
Currently in a conference, will do the change when I'm back.

**Comment by orbeckst** [comment]
RTD is happy again...

**Comment by xiki-tempula** [comment]
@orbeckst Thanks for the review. I have addressed all the versions and have changed the changes.

**Review by orbeckst — APPROVED** [review]

**Pull request merged**